README.md
# ArtColors: 
ArtColors provides intuitive color mixing in **subtractive Red-Yellow-Blue** color space.  It's a digital version of the traditional art school color wheel, which provides triadic, split-complementary,tetradic and analogous color harmonies.  It helps to select pleasing palettes of colors.  ArtColors also provides RGB-to-RYB conversion with simple formulas.  The program is built with RayLib and RayGui, but the RYB-RGB formulas can be used separately in your own code.

First beta release: March 2020.  Version 0.20 released in April 2021. See Roadmap.md and Changelog for details.

![ScreenShot](images/ArtColorsScreen.png)

![ScreenShot](images/ArtColorsScreenWhite.png)

# Goals
1. **Intuitive color mixing for graphic design, data visualization, GUI creation or video game graphics.**

   Everyone grows up learning that yellow and blue make green, not cyan!  Additive mixing in the RGB space is unintuitive.  Subtractive mixing in the RYB space is familiar.
   
2. **Easy generation of harmonious palettes for pleasing graphic design.**

   The basic color schemes in fine art are based on simple geometric relations on an RYB color wheel which has red, yellow and blue primaries 120 degrees apart.  The typical RGB color wheel has red, *green* and blue 120 degrees apart, so these relations fail.  In RGB, the blue-green spectrum tends to dominate, obscuring yellows, purples and browns.
   
3. **Portable simple formulas for RYB-to-RGB conversion and subtractive color mixing.**

   There is a lot of literature on subtractive color mixing.  My approach aims for a compromise between simplicity and accuracy, with much less code than a reflectance-based approach.  Most subtractive color mixing is done in the CMYK- space, rather than RYB, so hopefully these formulas fill a need.

# How To Use ArtColors
## To pick a color palette
Select a type of color palette (triadic, tetradic, etc.) by clicking on the Palette Type button, then rotate the Hue slider.  The small squares that move around the perimeter of the wheel indicate the principal colors of your palette.  They are the same colors as shown in the larger rectangles to the right of the wheel, which also show their RGB values.  The swatches to the right vary the hue and saturation of your principal colors to provide a wide range of options.  You can export these values to disk by pressing the "Save swatches to .pal" button.  Done!

You may also wish to check whether your palette is suitable for colorblind users.  In the lower right, clicking the "Colorblind mode" button changes your color swatches to approximate how they may be perceived by viewers with various color vision deficiencies: protanopia, deuteranopia, tritanopia and achromatopsia.  Click once more to return to normal view.  A colorblindness check is especially important for data visualization.

## To mix your own shades using RYB
On the lower left of the screen, there are two areas to generate colors from RYB values using sliders for Red, Yellow and Blue values.  These colors, or any other colors on the screen, can be blended by sampling them.  Click "Pick Color" and left-click on one color, then click "Pick Color" again and right-click on another color.  They will appear on the left and right sides of the color mixing bars in the middle-right of the screen.  The subtractive color mixing bar uses our algorithm to represent better how pigments blend in art media (paints, inks, pencils, crayons).  

## Explanation of the Interface
**Canvas Color** - Switches between black or white background.  You should definitely view your colors against both backgrounds, especially to appraise very light tints or dark shades.

**Palette Type** - Choose from one of six types of color harmonies (see explanation below).  The harmonies have different numbers of principal colors: Triadic has 3, Split Complementary has 3, Square Tetradic has 4, Rectangular Tetradic has 4, Complementary just 2, and Analogous has 4.  Your basic theme colors are those which appear to the right of the color wheel, and their locations are marked by small squares, similar to the windows in a handheld color wheel (see picture below).

**Palette Swatches** - Determined by your selected palette type.  A range of saturations and tints/shades based on the theme colors, from almost-black to almost-white, and from low saturation to high.

**Custom Color Mixing** - At the bottom of the screen are two Red-Yellow-Blue mixing areas.  Color A and Color B are mixed in the color mixing bars according to three kinds of mixing (see explanation below): Additive Quadratic mixing, Additive Linear mixing (the one you are used to), and *Subtractive* (yay!).  Check the tiny box above "Inverse" to display the *inverse color of the subtractive mix.*  Each color bar starts from 100% of Color A to 100% of Color B, ranging by 10% steps in between.

All Color Triplets shown on-screen are RGB values so you can copy them down if you'd like.

**Exporting** - Palettes are saved in .pal files, which are an ASCII readable standard.  You can import the palettes into GIMP, Paint Shop Pro, Adobe, etc.  *Save swatches* will save all the color swatches.  This will be 72 times the number of theme colors in your palette. *Save color bars* will save the color mixing bar of your choice.  Color bar palettes are only 11 colors in size.  If you check the "Inverse" box, an option to save the inverse color bar will also appear.

If you want to extract the RGB data from the .pal file with a text editor, the first three lines of the file are: a FOURCC code (JASC-PAL), a file type version (100) and the number of colors in the palette.  All subsequent numbers are RGB triplets for each color in the palette. Lines are terminated with \r\n.  The file can safely be edited to suit your needs.  Some programs may require either 16 or 256 colors in a .pal file; others don't abide by this convention, including GIMP.

**Screen Shot** - Useful if you want to import the whole screen into a graphics program, and sample there whatever hue you want.

# Challenges and Design Principles of ArtColors
A brief summary of the challenges of designing a subtractive Red-Yellow-Blue color space explains the principles I used to design ArtColors.  It also summarizes many of the things I learned while researching this project, and so I offer it in case it is instructive to others.

Hereafter common color space names will be abbreviated as follows:

**RYB-**=*Subtractive Red-Yellow-Blue color space*  Our color system.  
**RGB+**=*Additive Red-Green-Blue*  The system most widely used for monitor display. [Wiki](https://en.wikipedia.org/wiki/RGB_color_model)  
**HSV**=*Hue-Saturation-Value*  An alternative representation of the RGB+ system.  [Wiki](https://en.wikipedia.org/wiki/HSL_and_HSV)  
**CYMK-**=*Cyan-Yellow-Magenta-Black*  A subtractive system, most frequently used in printing.  [Wiki](https://en.wikipedia.org/wiki/CMYK_color_model)  

## Challenge #1: Everyone learns color mixing using RYB- primaries not RGB+
From the time you first played with crayons or fingerpaints, you've been using an RYB- color space.  Red plus green make a murky brown, but in RGB+ they make yellow.  Red and blue make a dark purple, but in RGB+ they make a bright magenta.  This can make it hard to quickly find the colors you want when blending.

RGB+ makes sense for what it was designed to do.  Computer monitors are emissive: they emit light which therefore blends additively.  No painting can do this.  Art media are reflective: the pigments we use absorb incident white light and reflect what they don't absorb.  Thus, they are subtractive: the more pigments one adds, the less light is reflected to the eye.  RGB+ was designed to align two of its primaries (blue and green) with two of the colors of light to which the three types of human [cone cells](https://en.wikipedia.org/wiki/Cone_cell) in the eye are most responsive.  RGB's third color, red, was chosen instead of the third cone color of yellow-green to give breadth of palette for mixing and the particular red was somewhat based on available emissive light sources at the time.<sup>[1](#footnote01)</sup>

Some disadvantage of RGB+ are: the vast majority our experience mixing colors is in a subtractive space, and many colors in our experience are non-spectral (see below).

Moreover, traditional instruction in art still uses RYB- to teach color theory.  Traditional art color theory has a long and complex history, in part driven by the range of pigments available to artists and their vast experience.  While the theoretical formalization of this body of knowledge into RYB- was premised in part (but only in part!) on erroneous 19th century ideas of the physics of color, the approach is still fundamentally valid -- as evidenced by the fact that artists, who devote enormous study to color -- still use it as a reference today.  The success of RYB- does not depend on its ability to theoretically unify the physics of light and human physiology, but on its *empirical validity* to accurately predict what color will result from subtractive color mixing.<sup>[2](#footnote02)</sup>

Thus almost everyone who has ever taken an art class or read a book on art color theory has encountered the standard RYB- color wheel:  
![Artist Color Wheel](images/ArtColorWheel.jpg)

### Principle #1: ArtColors should mimic a standard RYB- color wheel

## Challenge #2: RGB+ and HSV color wheels are blue-green dominated
Because two of the three primaries are *green and blue*, the yellow-orange range is severely compressed, purple is likewise narrow and dominated by magenta.  One has to be sniper to select these hues.  Small changes in RGB values or HSV hue quickly sweep through these ranges of color.

Compare the standard HSV color wheel with the ArtColor color wheel to see the problem and its solution:
![HSV Color Wheel](images/HSV-wheel.png) ![ArtColor Wheel](images/ArtColor-ColorWheel.png)

### Principle #2: ArtColors should make it easy to select a wide range of hues by balancing the spectrum
The ArtColor Wheel broadens the yellow-orange spectrum and adds deeper purples, while keeping distinct lines for cyan and magenta.  Note that red, yellow and blue primaries (small squares) are separated by 120 degrees.  This will faciliate traditional palette selection and blending (see below).  *How* to transform RGB+ to RYB- is where the math gets interesting.

If you are the inquisitive sort who wonders why RGB+ fails at providing a balanced spectrum of colors, there are two reasons.  The first we can summarize by analogy: green and blue aren't very "orthogonal" in our color experience.  But this merely begs the question: why not?  The answer is not trivial.  Indeed the question is even more perplexing when one considers that RGB is based in part on the physiology of human cone cells.  Shouldn't green and blue *seem* orthogonal to us insofar as they are aligned to the "vector bases" of our separate cone cell "inputs"?  The answer is that our *psychological* perception of color is not a direct result of human cone cell neuron response curves.  The brain *does its own mixing* to create colors we commonly experience but are not aligned to the frequency-response curve of any one cone cell.  Read about the Ewald Hering's opponent process theory of vision or the excellent overview in [Clive Maxfield's article](http://www.clivemaxfield.com/diycalculator/sp-cvision.shtml) for details.  Second, there are [non-spectral colors](https://en.wikipedia.org/wiki/Spectral_color#Non-spectral_colors): you don't see them in prismatically separated light.

Non-spectral colors include:  
* *All grays!* (thank you, rod cells)  
* *All tints and shades* = art terms for adding white or black respectively to a "pure" color.  All mixtures of a spectral color plus any shade of gray (from white to black) result in a non-spectral color.  In physical media, this happens all the time!  
* *Important hues like purple or brown* (and any tint or shade thereof, like pink or rose).  "Purple" isn't a spectral color: the perception of purple is the psychological result when the brain receives stimuli from *both* the red-sensitive and blue-sensitive cone cells at the same time.

So RGB is based on the physics of emitted light, and its primaries are selected (partly) based on the physiology of human cone cells.  But RGB *doesn't* factor in higher-level yet *intrinsically human* neurological factors of how the brain generates certain color perceptions that are non-spectral.  Art color theory was attentive to this fact from mere experience, even without understanding of the underlying causes.  In this regard, they were better empiricists, and RYB- theory reflects that advantage in my opinion.


## Challenge #3: Preserve traditional color harmonies in palette selection
The term [color harmony](https://blog.thepapermillstore.com/color-theory-color-harmonies/) is used in art to describe the theory of choosing attractive palettes which achieve different artistic effects.  In a *musical* harmony, notes played at the same time need to be sufficiently distinct to avoid dissonance, but the separation between the notes achieves the overall feeling of the chord (major chord vs. minor, etc.).  So likewise in a *color* harmony, colors are chosen that pair together to achieve a palette that produces a certain aesthetic effect.  (A bonus for *color* harmonization, unlike music, is that a narrow range of colors (called analogous) are not dissonant.)  A "complementary color" is its "opposite" on the traditional color wheel, which has its roots in the after-image one sees after looking at a bright color then looking at white.

A book on color theory for painting will effectively demonstrate how a painter's choice of palette greatly impacts his artwork.

On a traditional artist's color wheel, the theory of color harmonies is neatly summarized by the geometrical relationships seen in the central circle of the wheel:

![Artist Color Wheel](images/ArtColorWheel.jpg)

The traditional palette schemes and their geometries in RYB color space are:
* Triadic

   Colors separated by 120 degrees
* Split Complementary

   Two other colors offset by 30 degrees from the complementary color of the first.
   
* Square Tetradic

   Colors at 0, 90, 180, 270 degrees.  Two sets of complementary colors, separated by 90 degrees.
   
* Rectangular Tetradic

   Colors at 0, 120, 180, 300 degrees.  Pick two complementary colors.  For each, find their two triadic harmony colors and make these four a set.  These latter four colors make up the rectangular tetradic scheme.

**The problem: These geometrical relations depend on RYB primaries separated by 120 degrees, which RGB does not provide.**
Thus an RGB or HSV color wheel makes it difficult to select traditional harmonious palettes because green, not yellow, is 120 degrees from red and blue.  Even some sophisticated online color pickers claim to provide these color harmonies, but they fail because they use the correct degrees of separation, but the wrong basis (HSV), yielding incorrect colors.


CHANGELOG.md
## Current Release: Version 0.2, April 2021
Changes:
- Separated formulas for RYB conversion and subtractive mixing into their own header and source files (RYB.h & RYB.cpp).  They can now be used in stand-alone fashion either with or without Raylib
- Added, by request, a colorblind mode to facilitate palette selection for those with various color vision deficiencies (protanopia, deuteranopia, tritanopia or achromatopsia)
- Added color sampling functionality -- the "pick color" button
- Changed some font sizes for better legibility
- Simplified some of the code in formulas for better legibility (same functionality)
- Developed the README.md document further

## Original Release: Version 0.1, March 2020


ArtColors.cpp
//ArtColors by Eric J. Jenislawski
//Version 0.2
//Subtractive Red-Yellow-Blue color mixing and palette selector
//Copyright 2020, Eric J. Jenislawski.  Licensed under GNU General Public License version 3 or later.
//The extremely useful RayLib and RayGUI libraries are by Ramon Santamaria (raysan5) https://github.com/raysan5/


#include "raylib.h"
#include "raymath.h"
#define RAYGUI_IMPLEMENTATION
#include "raygui.h"
#include <iostream>
#include <cstdio> //for snprintf
#include <cmath> //for fabs
#include <vector> // for C++ vector
#include <fstream> //for saving palettes
#include <string> // for to_string() for saving palettes
#include <algorithm>
#include "RYB.h"

using namespace std;


//Full color from 300 degrees to 60 degrees.  No color on [120,240].  Smoothstep transition in between.
float step2(float deg) {
    float out=0.0;
    float sc=0.0;
    while (deg<0.0) { deg+=360.0;}
    while (deg>360.0) { deg-=360.0;}

    if (deg<=60.0) {
        out=1.0;
    }
    else if ( (deg>60.0)&&(deg<=120.0) ) {
        sc=(deg-60.0)/60.0;
        out=1.0-2.0*sc/sqrt(1.0+3.0*sc*sc);
    }
    else if ( (deg>120.0) && (deg<=240.0) ) {
        out=0.0;
    }
    else if ( (deg>240.0) && (deg<=300.0) ) {
        sc=(deg-240.0)/60.0;
        out=2.0*sc/sqrt(1.0+3.0*sc*sc);
    }
    else if ( (deg>300.0) && (deg<=360.0) ) {
        out=1.0;
    }

return out;
}

Color map2(float deg) {
    Vector3 out;
    Color output;

    //Function-based color spread around the wheel
    out.x=255*step2(deg);
    out.y=255*step2(deg-120);
    out.z=255*step2(deg-240);

    output=Xform_RYB2RGB(out.x,out.y,out.z);

return output;
}



// MAIN

int main()
{
//Initialize Raylib
    InitWindow(1500, 900, "ArtColors: RYB mixer and palette selector");
    SetWindowPosition(400,50);

    Camera2D camera = { 0 };
    camera.target = (Vector2){ 0, 0 };
    camera.offset = (Vector2){ 0, 0 };
    camera.rotation = 0.0f;
    camera.zoom = 1.0f;

    SetTargetFPS(30);

    Color col=BLUE;
    float degrees=0.0;
    float brightness=0.0;
    float sat=0.0;
    float myred=0.0,myyellow=0.0,myblue=0.0;
    float myred2=0.0,myyellow2=0.0,myblue2=0.0;
    char textout[20];
    Color BackgroundColor = BLACK;

    vector<float> PaletteDegrees;

    Vector2 CircleCenter={300,250};
    Vector2 MousePosition;

    Color mix_a, mix_b,pixcol_a,pixcol_b;
    bool mix_a_hook=false,mix_b_hook=false;

    bool HelpButtonState=false;
    int HelpBoxState = 0;
    bool TextExportButtonState=false;
    bool TextExportButtonState2=false;
    bool TextExportButtonState3=false;
    bool TextExportButtonState4=false;
    bool TextExportButtonState5=false;
    bool ScreenShotButtonState=false;
    bool DrawInverseColors=false;

    int PaletteSelectorResult = 0;
    int CanvasToggleGroup = 0;

    bool ColorPickerActive=false;
    int ColorblindMode=0;
    bool RenderForColorDeficiency=false;

    while (!WindowShouldClose()){

//Update

//TODO:
//Get something better for a color picker for any hue on the screen.
//Keep screen shot as method of export.
    if(ColorPickerActive) {
        if (IsMouseButtonPressed(MOUSE_LEFT_BUTTON)) {
            MousePosition=GetMousePosition();
            cout<<"Mouse X:"<<MousePosition.x<<" Y="<<MousePosition.y<<endl;
            Image screenshot=GetScreenData();
            Color* sscols=new Color[1500*900];
            sscols=GetImageData(screenshot);
            pixcol_a=sscols[1500*((int)MousePosition.y)+(int)MousePosition.x];
            printf("R=%i, G=%i, B=%i \n",pixcol_a.r,pixcol_a.g,pixcol_a.b);
            UnloadImage(screenshot);
            delete[] sscols;
            mix_a_hook=true;
            ColorPickerActive=false;
        }
        else if (IsMouseButtonPressed(MOUSE_RIGHT_BUTTON)) {
            MousePosition=GetMousePosition();
            cout<<"Mouse X:"<<MousePosition.x<<" Y="<<MousePosition.y<<endl;
            Image screenshot=GetScreenData();
            Color* sscols=new Color[1500*900];
            sscols=GetImageData(screenshot);
            pixcol_b=sscols[1500*((int)MousePosition.y)+(int)MousePosition.x];
            printf("R=%i, G=%i, B=%i \n",pixcol_b.r,pixcol_b.g,pixcol_b.b);
            UnloadImage(screenshot);
            delete[] sscols;
            mix_b_hook=true;
            ColorPickerActive=false;
        }
    }


//Draw

        BeginDrawing();
        ClearBackground(BackgroundColor);
        BeginMode2D(camera);

        DrawText("Canvas Color",10,30,16,GRAY);
        CanvasToggleGroup = GuiToggleGroup((Rectangle){ 10, 50, 50, 25 }, "BLACK;WHITE", CanvasToggleGroup);
        switch (CanvasToggleGroup)
        {
        case 0:
            BackgroundColor=BLACK;
            break;
        case 1:
            BackgroundColor=WHITE;
            break;
        default:
            BackgroundColor=BLACK;
        }

        for (int i=0;i<360;i+=2) {
            col=map2((float)i);
            col=Saturate(col,sat);
            col=Brighten(col,brightness);
            DrawCircleSector(CircleCenter,200,i,i+5,6,col);
            //DrawCircleSector({1000,350},200,i,i+5,6,(Color){RYBCol.x,RYBCol.y,RYBCol.z,255});
        }

        degrees=GuiSlider({150,500,360,20},NULL,NULL,degrees,0.0,359.0);
        snprintf(textout,sizeof textout,"%f",degrees);
        DrawText(textout,520,500,20,GRAY);
        DrawText("Hue:",100,500,20,GRAY);

        brightness=GuiSlider({150,550,360,20},NULL,NULL,brightness,-1.0,1.0);
        if (fabs(brightness)<0.05) {brightness=0.0;}
        snprintf(textout,sizeof textout,"%f",brightness);
        DrawText(textout,520,550,20,GRAY);
        DrawText("Tint/Shade:",20,550,20,GRAY);

        sat=GuiSlider({150,600,360,20},NULL,NULL,sat,-1.0,1.0);
        if (fabs(sat)<0.05) {sat=0.0;}
        snprintf(textout,sizeof textout,"%f",sat);
        DrawText(textout,520,600,20,GRAY);
        DrawText("Saturate:",40,600,20,GRAY);

        ColorPickerActive=GuiToggle({150,640,80,20},"Pick Color",ColorPickerActive);
        DrawText("Left click on Color A, right click on B",250,640,16,GRAY);

//Color Palette Selection
        DrawText("Palette Type",550,30,20,GRAY);
        PaletteSelectorResult = GuiComboBox((Rectangle){ 550, 50, 150, 30 }, "Triadic;SplitComp;SqTetrad;RectTetrad;Analogous;Compl", PaletteSelectorResult);
        PaletteDegrees.clear();
        switch (PaletteSelectorResult)
        {
        case 0:
            PaletteDegrees.push_back(degrees);
            PaletteDegrees.push_back(degrees+120.0);
            PaletteDegrees.push_back(degrees+240.0);
            break;
        case 1:
            PaletteDegrees.push_back(degrees);
            PaletteDegrees.push_back(degrees+150.0);
            PaletteDegrees.push_back(degrees+210.0);
            break;
        case 2:
            PaletteDegrees.push_back(degrees);
            PaletteDegrees.push_back(degrees+90.0);
            PaletteDegrees.push_back(degrees+180.0);
            PaletteDegrees.push_back(degrees+270.0);
            break;
        case 3:
            PaletteDegrees.push_back(degrees);
            PaletteDegrees.push_back(degrees+120.0);
            PaletteDegrees.push_back(degrees+180.0);
            PaletteDegrees.push_back(degrees+300.0);
            break;
        case 4:
            PaletteDegrees.push_back(degrees);
            PaletteDegrees.push_back(degrees+15.0);
            PaletteDegrees.push_back(degrees+30.0);
            PaletteDegrees.push_back(degrees+45.0);
            break;
        case 5:
            PaletteDegrees.push_back(degrees);
            PaletteDegrees.push_back(degrees+180.0);
            break;
        }

//Draw Indicators Around Color Wheel
        for (unsigned int s=0;s<PaletteDegrees.size();s++) {
            float deg=PaletteDegrees[s];
            col=map2(deg);
            col=Saturate(col,sat);
            col=Brighten(col,brightness);
            DrawRectangle(CircleCenter.x+220*cos(DEG2RAD*(-deg+90)),CircleCenter.y+220*sin(DEG2RAD*(-deg+90)),10,10,col);
            DrawRectangle(550,200+30*s,40,20,col);
            snprintf(textout,sizeof textout,"%i %i %i",col.r,col.g,col.b);
            DrawText(textout,600,200+30*s,20,GRAY);

        }

/*
        //Alternative Color Picker
        float picker=GuiSlider({950,50,510,20},NULL,NULL,picker,0,255);
        for (int i=0;i<255;i++) {
            for (int j=0;j<255;j++) {
                col=Xform_RYB2RGB((int)picker,i,j);
                col=Saturate(col,sat);
                col=Brighten(col,brightness);
                DrawRectangle(1000+i,100+j,5,5,col);
            }
        }
*/


//Palettizer

        GuiGroupBox((Rectangle){ 800,150,615,215 }, "Swatches");
        for (unsigned int s=0;s<PaletteDegrees.size();s++) {
            col=map2(PaletteDegrees[s]);
            Color PCol;
            int sep=60/PaletteDegrees.size();

            for (float i=-0.8;i<1.0;i+=0.2) {
                for (float j=-0.8;j<=0.8;j+=0.2) {
                    PCol=Saturate(col,i);
                    PCol=Brighten(PCol,j);
                    if (RenderForColorDeficiency) {
                        PCol=ColorBlindTransform(PCol,ColorblindMode);
                    }
                    DrawRectangle(1050+i*300+sep*s,250+j*100,15,15,PCol);
                }
            }
        }


/*
    // RYB-RGB function and inverse test pattern
    //A round-trip conversion from RGB to RYB to RGB.  Helps show where trilinear interpretation values can be improved
    //Uncomment this section to draw a color bar beneath the save buttons and above the color swatches

    for (float i=0.0;i<360;i+=10.0) {
        col=map2(i);
        DrawRectangle(700+2*i,90,20,20,col);
        col=Xform_RGB2RYB(col.r,col.g,col.b);
        col=Xform_RYB2RGB(col.r,col.g,col.b);
        col=Saturate(col,0.5);
        DrawRectangle(700+2*i,110,20,20,col);
    }
*/

//Custom blender

        DrawText("Custom Color Blender",100,750,20,GRAY);

        myred=GuiSlider({100,780,255,20},"Red",NULL,myred,0.0,255.0);
        col=Xform_RYB2RGB(myred,0,0);
        DrawRectangle(360,780,20,20,col);

        myyellow=GuiSlider({100,810,255,20},"Yellow",NULL,myyellow,0.0,255.0);
        col=Xform_RYB2RGB(0,myyellow,0);
        DrawRectangle(360,810,20,20,col);

        myblue=GuiSlider({100,840,255,20},"Blue",NULL,myblue,0.0,255.0);
        col=Xform_RYB2RGB(0,0,myblue);
        DrawRectangle(360,840,20,20,col);

        col=Xform_RYB2RGB(myred,myyellow,myblue);
        DrawRectangle(400,780,100,100,col);

        snprintf(textout,sizeof textout,"%i %i %i",col.r,col.g,col.b);
        DrawText(textout,520,780,20,GRAY);
//        mix_a=col;
        if (mix_a_hook==true) {
            mix_a=pixcol_a;
//            mix_a_hook=false;
        }
        else {
            mix_a=col;
        }


        myred2=GuiSlider({700,780,255,20},"Red",NULL,myred2,0.0,255.0);
        col=Xform_RYB2RGB(myred2,0,0);
        DrawRectangle(960,780,20,20,col);

        myyellow2=GuiSlider({700,810,255,20},"Yellow",NULL,myyellow2,0.0,255.0);
        col=Xform_RYB2RGB(0,myyellow2,0);
        DrawRectangle(960,810,20,20,col);

        myblue2=GuiSlider({700,840,255,20},"Blue",NULL,myblue2,0.0,255.0);
        col=Xform_RYB2RGB(0,0,myblue2);
        DrawRectangle(960,840,20,20,col);

        col=Xform_RYB2RGB(myred2,myyellow2,myblue2);
        DrawRectangle(1000,780,100,100,col);

        snprintf(textout,sizeof textout,"%i %i %i",col.r,col.g,col.b);
        DrawText(textout,1120,780,20,GRAY);

//        mix_b=col;
        if (mix_b_hook==true) {
            mix_b=pixcol_b;
//            mix_b_hook=false;
        }
        else {
            mix_b=col;
        }


        DrawText("Additive",700,410,16,GRAY);
        DrawText("Quadratic:",700,430,16,GRAY);

        DrawText("Additive",700,480,16,GRAY);
        DrawText("Linear:",700,500,16,GRAY);

        DrawText("Subtractive:",700,560,16,GRAY);

        DrawInverseColors=GuiToggle((Rectangle){700,620,10,10}," ",DrawInverseColors);
        DrawText("Inverse:",700,640,16,GRAY);

        for (float i=0.0;i<1.1;i+=0.1) {
            col=ColorMix(mix_a,mix_b,i);
            DrawRectangle(800+(600*i),400,60,50,col);
            snprintf(textout,sizeof textout,"%i %i %i",col.r,col.g,col.b);
            DrawText(textout,800+(600*i),450,10,GRAY);

            col=ColorMixLin(mix_a,mix_b,i);
            DrawRectangle(800+(600*i),470,60,50,col);
            snprintf(textout,sizeof textout,"%i %i %i",col.r,col.g,col.b);
            DrawText(textout,800+(600*i),520,10,GRAY);

            col=ColorMixSub(mix_a,mix_b,i);
            DrawRectangle(800+(600*i),540,60,50,col);
            snprintf(textout,sizeof textout,"%i %i %i",col.r,col.g,col.b);
            DrawText(textout,800+(600*i),590,10,GRAY);

            if (DrawInverseColors) {
                col=ColorInv(col);
                DrawRectangle(800+(600*i),620,60,30,col);
                snprintf(textout,sizeof textout,"%i %i %i",col.r,col.g,col.b);
                DrawText(textout,800+(600*i),660,10,GRAY);
            }
        }

//Export and Help Buttons
        TextExportButtonState=GuiButton((Rectangle){800,50,130,35},"Save Swatches to .pal");
        if (TextExportButtonState) {
            ofstream PaletteFile;
            PaletteFile.open("ArtColor-Swatches.pal");
            if (PaletteFile.is_open()) {
                PaletteFile<<"JASC-PAL\r\n";
                PaletteFile<<"0100\r\n";
                PaletteFile<<to_string(PaletteDegrees.size()+72*PaletteDegrees.size())<<"\r\n";
                for (unsigned int s=0;s<PaletteDegrees.size();s++) {     //This part should match Palettizer loops.  TODO: make a subroutine for both.
                    col=map2(PaletteDegrees[s]);
                    Color PCol;
                    for (float i=-0.8;i<1.0;i+=0.2) {
                        for (float j=-0.8;j<=0.8;j+=0.2) {
                            PCol=Saturate(col,i);
                            PCol=Brighten(PCol,j);
                            PaletteFile<<to_string(PCol.r)<<" "<<to_string(PCol.g)<<" "<<to_string(PCol.b)<<"\r\n";
                        }
                    }
                }
                PaletteFile.close();
                cout<<"Palette file saved."<<endl;
            }
            else {
                cout<<"File creation failed."<<endl;
            }
        }

        DrawText("Save Color Bars:",950,30,20,GRAY);
        TextExportButtonState2=GuiButton((Rectangle){950,50,50,35},"Quad+");
        if (TextExportButtonState2) {
            ofstream PaletteFile;
            PaletteFile.open("ArtColor-AddQuadBars.pal");
            if (PaletteFile.is_open()) {
                PaletteFile<<"JASC-PAL\r\n";
                PaletteFile<<"0100\r\n";
                PaletteFile<<to_string(11)<<"\r\n";
                for (float i=0.0;i<1.1;i+=0.1) {
                    col=ColorMix(mix_a,mix_b,i);
                    PaletteFile<<to_string(col.r)<<" "<<to_string(col.g)<<" "<<to_string(col.b)<<"\r\n";
                }
                PaletteFile.close();
                cout<<"Additive Quadratic bars palette saved."<<endl;
            }
            else {
                cout<<"File creation failed."<<endl;
            }
        }

        TextExportButtonState3=GuiButton((Rectangle){1020,50,50,35},"Lin+");
        if (TextExportButtonState3) {
            ofstream PaletteFile;
            PaletteFile.open("ArtColor-AddLinBars.pal");
            if (PaletteFile.is_open()) {
                PaletteFile<<"JASC-PAL\r\n";
                PaletteFile<<"0100\r\n";
                PaletteFile<<to_string(11)<<"\r\n";
                for (float i=0.0;i<1.1;i+=0.1) {
                    col=ColorMixLin(mix_a,mix_b,i);
                    PaletteFile<<to_string(col.r)<<" "<<to_string(col.g)<<" "<<to_string(col.b)<<"\r\n";
                }
                PaletteFile.close();
                cout<<"Additive Linear bars palette saved."<<endl;
            }
            else {
                cout<<"File creation failed."<<endl;
            }
        }

        TextExportButtonState4=GuiButton((Rectangle){1090,50,50,35},"Subtract");
        if (TextExportButtonState4) {
            ofstream PaletteFile;
            PaletteFile.open("ArtColor-SubBars.pal");
            if (PaletteFile.is_open()) {
                PaletteFile<<"JASC-PAL\r\n";
                PaletteFile<<"0100\r\n";
                PaletteFile<<to_string(11)<<"\r\n";
                for (float i=0.0;i<1.1;i+=0.1) {
                    col=ColorMixSub(mix_a,mix_b,i);
                    PaletteFile<<to_string(col.r)<<" "<<to_string(col.g)<<" "<<to_string(col.b)<<"\r\n";
                }
                PaletteFile.close();
                cout<<"Subtractive mixing bars palette saved."<<endl;
            }
            else {
                cout<<"File creation failed."<<endl;
            }
        }

        if (DrawInverseColors) {
        TextExportButtonState5=GuiButton((Rectangle){1160,50,50,35},"Inv. Sub");
        if (TextExportButtonState5) {
            ofstream PaletteFile;
            PaletteFile.open("ArtColor-Inverse.pal");
            if (PaletteFile.is_open()) {
                PaletteFile<<"JASC-PAL\r\n";
                PaletteFile<<"0100\r\n";
                PaletteFile<<to_string(11)<<"\r\n";
                for (float i=0.0;i<1.1;i+=0.1) {
                    col=ColorMixSub(mix_a,mix_b,i);
                    col=ColorInv(col);
                    PaletteFile<<to_string(col.r)<<" "<<to_string(col.g)<<" "<<to_string(col.b)<<"\r\n";
                }
                PaletteFile.close();
                cout<<"Inverse bars palette saved."<<endl;
            }
            else {
                cout<<"File creation failed."<<endl;
            }
        }
        }


        DrawText("Colorblind mode:",1240,780,20,GRAY);
        ColorblindMode = GuiComboBox((Rectangle){ 1240, 820, 150, 35 }, "Normal;Protanopia;Deuteranopia;Tritanopia;Achromatopsia", ColorblindMode);
        if (ColorblindMode==0) RenderForColorDeficiency=false; else RenderForColorDeficiency=true;


        ScreenShotButtonState=GuiButton((Rectangle){1300,50,70,35},"Screen Shot");



        HelpButtonState=GuiToggle((Rectangle){1380,50,50,35},"Help",HelpButtonState);
        if (HelpButtonState) {
            string helpmessage="Choose a palette by moving the Hue / Brightness / Saturation sliders and selecting your palette type by cycling clicking the Palette Type button.\n";
            helpmessage+="This will update the color swatches on the right which span a range of tints and saturations.\n";
            helpmessage+="You can blend two colors by selecting them using the Pick Color button.  Click the button then click anywhere in the program window to sample a color, i.e., eyedropper mode\n";
            helpmessage+="Left click selects the first color for blending and right click selects the second color.  The mixing bars show different ways of mixing the two colors.  Subtractive uses the ArtColors formula.\n";
            helpmessage+="You can take the colors from the color wheel or use the RYB blending areas below to specify your own colors then select them to mix them.\n";
            helpmessage+="The save buttons above the swatches will export palette values for the swatches or for the mixing bars and inverse color bar respectively.\n";
            helpmessage+="The Colorblind Mode selector (lower right) replicates how the palette in the swatches would appear to people with various color vision deficiencies.\n";
            helpmessage+="You can use it to check, for example, whether your choice of palette colors are visually distinct to the colorblind.\n";
            helpmessage+="All triplets of numbers are RGB values.  Pressing ESC quits the program. \n";
            HelpBoxState=GuiMessageBox((Rectangle){200,400,1000,400},"Help Info",helpmessage.c_str(),"OK");
            if (HelpBoxState==0||HelpBoxState==1) HelpButtonState=false;
        }


        EndMode2D();
 //       DrawFPS(10,10);
        EndDrawing();

        if (ScreenShotButtonState) {
            TakeScreenshot("ArtColors-ScreenShot.png");
        }

    }

    return 0;
}
